timely_beliefs/beliefs/utils.py

```diff
@@ -1030,3 +1036,66 @@ def meta_repr(
             for attr in tb_structure._metadata
         ]
     )
+
+
+def upsample_beliefs_data_frame(
+    df: "classes.BeliefsDataFrame" | pd.DataFrame, event_resolution: timedelta, keep_nan_values: bool = False
+) -> "classes.BeliefsDataFrame":
+    """Because simply doing df.resample().ffill() does not correctly resample the last event in the data frame.
+
+    :param df:                  In case of a regular pd.DataFrame, make sure to set df.event_resolution before passing it to this function.
+    :param event_resolution:    Resolution to upsample to.
+    :param keep_nan_values:     If True, place back resampled NaN values. Drops NaN values by default.
+    """
+    from_event_resolution = df.event_resolution
+    resample_ratio = pd.to_timedelta(
+        to_offset(from_event_resolution)
+    ) / pd.Timedelta(event_resolution)
+    if keep_nan_values:
+        # back up NaN values
+        unique_event_value_not_in_df = df["event_value"].abs().sum() + 1
```

Comment:
Can you add a comment here saying that we are flagging the positions of the NaN values from the original data with a unique number? Also, an example on how this number is generated. 

Aside, what if df was a DataFrame full of NaN, wouldn't `unique_event_value_not_in_df` be NaN?  In this case it doesn't matter as we would keep the NaN.

Reply:
> Aside, what if df was a DataFrame full of NaN, wouldn't `unique_event_value_not_in_df` be NaN?  In this case it doesn't matter as we would keep the NaN.

It looks like this would also behave as we want it to. A unique number not already present in the series is created, as we can see here:
```
import numpy as np
import pandas as pd

s = pd.Series([np.NaN, np.NaN])
L1norm = s.abs().sum()
print(L1norm)  # 0
```